### PR TITLE
Track line execution counts

### PR DESCRIPTION
**Pull request by Badbond**
**Note:** This description is quite outdated due to more recent changes. Please take a look at the branches and tags below for implementations of tracking line execution counts. The git tag also contains additional documentation on its changes and perceived impacts. The branches are based on `v0.8.7`.

- Using overflowing integers: [branch](https://github.com/Badbond/jacoco/tree/track-line-execution-count-integer-overflow), [tag](https://github.com/Badbond/jacoco/releases/tag/integer-overflow).
- Using overflowing longs: [branch](https://github.com/Badbond/jacoco/tree/track-line-execution-count-long-overflow), [tag](https://github.com/Badbond/jacoco/releases/tag/long-overflow).
- Using `Math.min` to cap at `Integer.MAX_VALUE - 1`: [branch](https://github.com/Badbond/jacoco/tree/track-line-execution-count-math-min), [tag](https://github.com/Badbond/jacoco/releases/tag/math-min).
- Using double casting to cap at `Integer.MAX_VALUE`: [branch](https://github.com/Badbond/jacoco/tree/track-line-execution-count-double-cast), [tag](https://github.com/Badbond/jacoco/releases/tag/double-cast).
- Using BigInteger (WIP -- abandoned): [branch](https://github.com/Badbond/jacoco/tree/track-line-execution-count-big-int-wip).

---

## Description

This PR attempts to add the **experimental** functionality to count how often a line was executed by converting the `boolean` probe array to an `int` probe array. This is to be used in a thesis project as part of dynamic analysis if we can pull it off with not too much overhead.  

Noteworthy changes that a review would need to focus on:

- The `ExecutionData` now holds an `int[]` instead of a `boolean[]`. 
- Changes to the `ProbeInserter` as we now insert different bytecode.
- Changes to `Instruction` on how we map the probe execution counts to the instruction level and on how we handle instruction propagation, branching, branch merging and branch replacement. 
- Use of execution data in `LineImpl`. Here we effectively take the `max` of executions counts on all instructions on that line. 
- Use of execution data from `ILine` in reporting.

## Performance

Performance testing for the integer-with-overflow version shows a statistically significant (1 tailed unequal variance t-test, n=20, p < 0.01) decrease in performance during running a Maven build with tests of a private repository using `jacoco:prepare-agent` and `jacoco:report`.  This is as expected as we perform more instructions in bytecode. The arithmetic mean of the build times increased from 6m3s641ms to 6m8s862ms (+1.4%).  

### Implement using number without overflow

See #4 

We could use `BigInteger` to store the execution counts. This will require more memory to be used and might perform worse than primitive types. The bytecode would be something similar as the following:

```
    ALOAD <field-number>
    LDC <index-number>
    DUP2
    AALOAD
    GETSTATIC java/math/BigInteger.ONE : Ljava/math/BigInteger;
    INVOKEVIRTUAL java/math/BigInteger.add (Ljava/math/BigInteger;)Ljava/math/BigInteger;
    AASTORE
```

### Implement with a larger data type with overflow

See #3 

We could also use the `long` data type to allow for a bigger max value. We are then less likely to hit overflow here if we can limit our instrumentation duration or can reset it in such a way that we can only receive data for a limited amount of time after reset such that the chance of hitting overflow will be negligible. This would result in:

```
    ALOAD <field-number>
    LDC <index-number>
    DUP2
    LALOAD
    LCONST_1
    LADD
    LASTORE
``` 

The challenge with this implementation will be determining the max stack size. Ints take up one in stack size (32-bit), longs take up two (64 bit). However, BigInt can be many more. 

### Cap at max value

We could also cap the value in the data type to its max size (also applicable when choosing longs). 

#### Using Math library

See #2 

Creating frames is difficult as we don't know what the stack and locals are when inserting the probe. Therefore, these need to be computed and that is costly in performance and requires a lot of effort to change in JaCoCo. Instead, let's make the JVM handle that for us by using the static library `Math`. The downside is that this is probably **way** worse in performance compared to a simple if-check. It would be similar to:

```
    ALOAD <field-number>
    LDC <index-number>
    DUP2
    IALOAD
    ICONST_1
    IADD
    LDC 2147483646
    INVOKESTATIC java/lang/Math.min (II)
    IASTORE
```

#### Using an if-check

Perhaps there is a more efficient way than the compiled if-check for this. Such implementation in bytecode can be seen below:

```
    ALOAD <field-number>
    LDC <index-number>
    IALOAD
    LDC 2147483647
    IF_ICMPEQ <+7>
    ALOAD <field-number>
    ICONST_<index-number>
    DUP2
    IALOAD
    ICONST_1
    IADD
    IASTORE
```

The challenge with this implementation is that we do branching and therefore need to specify the stack map table in a frame. Unfortunately, [as the JaCoCo authors mention](https://github.com/jacoco/jacoco/pull/177#issue-71739831), we can't determine the stack before the probe insertion, which is needed to set it to the same stack after the probe:

> JaCoCo cannot re-calculate stackmap frames as this is an expensive operation and would require knowing the type hierarchy (which is not available). Also ASM default implementation for the type hierarchy actually loads(!) the classes.

Therefore, this approach is invalid.

### CODE REVIEW

File: org.jacoco.agent.rt.test/src/org/jacoco/agent/rt/internal/AgentTest.java [hunk 1]
```diff
@@ -266,14 +266,14 @@ public void setSessionId_should_modify_session_id() throws Exception {
...
+		int[] probes = agent.getData()
 				.getExecutionData(Long.valueOf(0x12345678), "Foo", 1)
 				.getProbes();
-		probes[0] = true;
+		probes[0] = 1;
 
 		agent.reset();
 
-		assertFalse(probes[0]);
+		assertFalse(probes[0] > 0);
```
Comment (marchof) [inline org.jacoco.agent.rt.test/src/org/jacoco/agent/rt/internal/AgentTest.java]: If the reset actually works the new value should 0.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/Instruction.java [hunk 2]
```diff
@@ -108,26 +113,34 @@ public void addBranch(final Instruction target, final int branch) {
...
+			int count) {
 		// No recursion here, as there can be very long chains of instructions
 		while (insn != null) {
 			if (!insn.coveredBranches.isEmpty()) {
-				insn.coveredBranches.set(branch);
+				// As propagation happens multiple times in the chain, we can
+				// not just increment the execution count. Instead, we need to
+				// set it to the max of the value that we already have or the
+				// value that is coming in from upstream. Otherwise, we will
+				// count more executions than actually happened.
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/Instruction.java:137]: Nice 👍

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/SourceNodeImpl.java [hunk 3]
```diff
@@ -107,27 +107,30 @@ public void increment(final ISourceNode child) {
...
+	 * @param executions
+	 *            executions to add
 	 * @param branches
 	 *            branches to add
 	 * @param line
 	 *            optional line number or {@link ISourceNode#UNKNOWN_LINE}
 	 */
-	public void increment(final ICounter instructions, final ICounter branches,
-			final int line) {
+	public void increment(final ICounter instructions, final int executions,
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/SourceNodeImpl.java:117]: I think we call this method for every bytecode instruction. So the new executions counter is the total number of all executions of all instruction, and not the number of time the line was executed. Is that intended?

File: org.jacoco.core/src/org/jacoco/core/analysis/ILine.java [hunk 4]
```diff
@@ -45,4 +45,10 @@
 	 */
 	int getStatus();
 
+	/**
+	 * Returns the execution count for this line
+	 *
+	 * @return execution count
+	 */
+	int getExecutionCount();
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/analysis/ILine.java:53]: I think this definition does not reflect the actual implementation. See my other comment below.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/SourceNodeImpl.java (hunk 3 above)
Comment (Badbond) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/SourceNodeImpl.java:117]: Ah, yes. For my thesis I'm only interested in how often a line was executed. The quickest way to implement this was to propagate it along with these functions all the way down to `LineImpl#increment`. So in that regard, the Javadoc is a bit off here. I don't store the total number of executions on all instructions on a class/method level (but I could). 

For line executions, I don't sum all the instruction's executions but rather take the `max` again (see `LineImpl#increment` implementations). Summing would result in an indication that a line was executed more than it actually was. However, it also might give a wrong perception about how often specific source code was executed as some statements might be executed less than reported (which I will carefully document in my thesis :smile:)

**Comment by Badbond** [comment]
Closing as part of Fork clean up. See branch [track-line-execution-count-integer-overflow](https://github.com/Badbond/jacoco/tree/track-line-execution-count-integer-overflow) and its tag [integer-overflow](https://github.com/Badbond/jacoco/releases/tag/integer-overflow) for this functionality.

Thank you Mark for taking a look at the code, it helped a lot to clarify some of the context of the inner workings of JaCoCo.

**Pull request closed without merging**